clash-prelude/src/Clash/Class/HasDomain/HasSingleDomain.hs

```diff
@@ -0,0 +1,179 @@
+{-|
+Copyright  :  (C) 2019, Myrtle Software Ltd
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  Christiaan Baaij <christiaan.baaij@gmail.com>
+-}
+
+{-# LANGUAGE CPP                  #-}
+{-# LANGUAGE ConstraintKinds      #-}
+{-# LANGUAGE PolyKinds            #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE TemplateHaskell      #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+#if __GLASGOW_HASKELL__ < 806
+{-# LANGUAGE TypeInType #-}
+#endif
+
+{-# OPTIONS_GHC -Wno-missing-methods #-}
+
+module Clash.Class.HasDomain.HasSingleDomain where
+
+import           Clash.Class.HasDomain.Common
+import           Clash.Class.HasDomain.CodeGen    (mkTryDomainTuples)
+
+import           Clash.Sized.Vector               (Vec)
+import           Clash.Sized.RTree                (RTree)
+import           Clash.Signal.Internal
+  (Signal, Domain, Clock, Reset, Enable)
+import           Clash.Signal.Delayed.Internal    (DSignal)
+
+import           Data.Kind                        (Type)
+import           Data.Proxy                       (Proxy)
+import           Type.Errors
+  (DelayError, TypeError, IfStuck, Pure)
+
+type MissingInstance =
+        "This might happen if an instance for TryDomain is missing. Try to determine"
+  :$$$: "which of the types miss an instance, and add them. Example implementations:"
+  :$$$: ""
+  :$$$: " * type instance HasDomain t (MyVector n a)    = TryDomain t a"
+  :$$$: " * type instance HasDomain t (MyCircuit dom a) = Found dom"
+  :$$$: " * type instance HasDomain t Terminal          = NotFound"
+  :$$$: ""
+  :$$$: "Alternatively, use one of the withSpecific* functions."
+
+type Outro =
+         ""
+   :$$$: "------"
+   :$$$: ""
+   :$$$: "You tried to apply an explicitely routed clock, reset, or enable line"
+   :$$$: "to a construct with, possibly, an implicitely routed one. Clash failed to"
+   :$$$: "unambigously determine a single domain and could therefore not route it."
```

Comment:
```suggestion
   :$$$: "unambiguously determine a single domain and could therefore not route it."
```

Reply:
And my editor even highlights spelling mistakes :sweat: . Thanks!